pkg/kubelet/images/image_manager.go

```diff
@@ -254,9 +238,47 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
 		return "", msg, err
 	}
 
+	pullCredentials, err := lookupPullCredentials()
+	if err != nil {
+		return "", err.Error(), err
+	}
+
 	return m.pullImage(ctx, logPrefix, objRef, pod.UID, requestedImage, spec, pullCredentials, podSandboxConfig)
 }
 
+func (m *imageManager) makeLookupPullCredentialsFunc(image string, pod *v1.Pod, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) func() ([]credentialprovider.TrackedAuthConfig, error) {
+	return sync.OnceValues(func() ([]credentialprovider.TrackedAuthConfig, error) {
```

Comment:
why do we need to wrap it in OnceValues? It is only called once anyways as far as I can tell

Reply:
the return lookup function might be called twice in some cases:
1. JIT during the MustAttemptImagePull
2. just before calling `m.pullImage()`

Reply:
> Will it retry if the first attempt resulted in `err`?

No. It caches both values.

> Or we want to cache the error

An error in that place shows that something is seriously misconfigured, so it is unlikely that retrying makes sense. A mere failure to lookup credentials from the keyring does not return an error.

Reply:
> A mere failure to lookup credentials from the keyring does not return an error.

Not because of this PR, I mean. The relevant method's signature does not include `error`.
```go
// pkg/credentialprovider/keyring.go
type DockerKeyring interface {
	Lookup(image string) ([]TrackedAuthConfig, bool)
}
```

> I plan to take it again later during 1.35, please ping if you think it is critical to do earleir

I think this change should be in place when `KubeletEnsureSecretPulledImages` feature gate goes GA and default to enabled, because at that point it will become enabled by some managed (cloud) Kubernetes providers with no option to disable.

Reply:
> An error in that place shows that something is seriously misconfigured, so it is unlikely that retrying makes sense. A mere failure to lookup credentials from the keyring does not return an error.

Besides, I would like to make sure the result of the credential lookup matches what the authz check received in the context of a single `EnsureImageExists()` call.

> I will need to review the sequence you suggested. I plan to take it again later during 1.35, please ping if you think it is critical to do earlier

This likely prevents us from moving the feature to Beta. It's not critical right now, but let's please make sure we pay attention to this PR early in 1.35.